assets/js/wc-gateway-ppec-generate-cart.js

```diff
@@ -30,52 +30,67 @@
 			$( '#woo_pp_ec_button_product > *' ).css( 'pointer-events', 'none' );
 		} );
 
+	var variation_valid = true;
+	var fields_valid = true;
+	var update_button = function() {
+		$( '#woo_pp_ec_button_product' ).trigger( ( variation_valid && fields_valid ) ? 'enable' : 'disable' );
+	};
+
+	var validate_form = function() {
+		silent_validation = true;
+		fields_valid = true;
+		form.find( 'select:enabled, input:enabled, textarea:enabled' ).each( function() {
+			fields_valid = fields_valid && this.checkValidity();
+		} );
+		silent_validation = false;
+		update_button();
+	};
+
 	// It's a variations form, button availability should depend on its events
 	if ( $( '.variations_form' ).length ) {
-		$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+		variation_valid = false;
 
 		$( '.variations_form' )
 		.on( 'show_variation', function( event, form, purchasable ) {
-			$( '#woo_pp_ec_button_product' ).trigger( purchasable ? 'enable' : 'disable' );
+			variation_valid = purchasable;
+			update_button();
 		} )
 		.on( 'hide_variation', function() {
-			$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+			variation_valid = false;
+			update_button();
 		} );
 	}
 
-	var get_attributes = function() {
-		var select = $( '.variations_form' ).find( '.variations select' ),
-			data   = {},
-			count  = 0,
-			chosen = 0;
-
-		select.each( function() {
-			var attribute_name = $( this ).data( 'attribute_name' ) || $( this ).attr( 'name' );
-			var value	  = $( this ).val() || '';
-
-			if ( value.length > 0 ) {
-				chosen++;
-			}
-
-			count++;
-			data[ attribute_name ] = value;
-		} );
-
-		return {
-			'count'      : count,
-			'chosenCount': chosen,
-			'data'       : data
-		};
-	};
+	// Disable the button if there are invalid fields in the product page (like required fields from Product Addons)
+	var silent_validation;
+	var form = $( 'form.cart' );
+	form.get( 0 ).addEventListener( 'invalid', function( e ) {
+		if ( silent_validation ) {
+			e.preventDefault();
+		}
+	}, true );
+	form.on( 'change', 'select, input, textarea', validate_form );
+	validate_form();
 
 	var generate_cart = function( callback ) {
 		var data = {
-			'nonce':       wc_ppec_generate_cart_context.generate_cart_nonce,
-			'qty':         $( '.quantity .qty' ).val(),
-			'attributes':  $( '.variations_form' ).length ? get_attributes().data : [],
-			'add-to-cart': $( '[name=add-to-cart]' ).val(),
+			'nonce': wc_ppec_generate_cart_context.generate_cart_nonce,
 		};
 
+		var field_pairs = form.serializeArray();
```

Comment:
I'm not sure I understand why we replace `get_attributes()` with the loop below. With `get_attributes` we cared only about `.variations_form` and not the entire form. 

Can you explain what `data` we expected before, what we now expect so see in `data`, and why the changes are needed?

Reply:
As you pointed out, `get_attributes()` only cared about the variations fields. By serializing the whole form, you get those variation fields **and** any fields added by a plugin **and** the basic fields like `quantity` and `add-to-cart` (product ID).

This logic is simpler, and it basically sends on a POST request the same data that would be sent in a plain HTML form (if JS was disabled). Product Add-Ons relies on that assumption, for example (it fetches the values directly from `$_POST` whenever `WC()->cart->add_to_cart()` is called).

That's why I merged the 2 issues into this 1 PR, because the fi for both variations misbehaving and Product Add-ons compatibility is basically the same: send all the HTML form data to the `wc_ajax_generate_cart()` AJAX endpoint.

Reply:
Thanks for explaining! I got it now 😄